### ISSUE TITLE
FromArray deriving issue when field type is array of objects

### ISSUE THREAD

**Issue by MassiAtHG** [prolic/fpp#103]
The following definition:
```
namespace HGEntity {
    data User : EntityInterface = User { 
        \HGEntity\VO\Generic\Id $id, 
        \HGEntity\VO\Generic\Name $fname,
        \HGEntity\VO\Generic\Name $lname,
        \HGEntity\VO\Generic\EmailAddress[] $email,
        \HGEntity\VO\Generic\Phone[] $mobile,
        \HGEntity\VO\Generic\Phone[] $phone,
        \HGEntity\VO\Generic\DateOfBirth $dob,
        \HGEntity\VO\User\Location $loc,
        \HGEntity\VO\Generic\Group $group,
        ?\HGEntity\VO\User\AvatarId $avatar
    } deriving (FromArray, ToArray, Equals) where 
        User:
            | strlen($fname->toString()) < 2    => 'First name too short!'
            | strlen($lname->toString()) < 2    => 'Last name too short!'
            | empty($this->verifyEmail($email)) => 'Invalid email provided!'
    ;
}
```
generates class code with function 'fromArray' containing checks of \is_HGEntity\VO\Generic\Phone or \is_HGEntity\VO\Generic\EmailAddress as follow:
```
public static function fromArray(array $data): User
    {
        if (! isset($data['id']) || ! \is_string($data['id'])) {
            throw new \InvalidArgumentException("Key 'id' is missing in data array or is not a string");
        }

        $id = \HGEntity\VO\Generic\Id::fromString($data['id']);

        if (! isset($data['fname']) || ! \is_string($data['fname'])) {
            throw new \InvalidArgumentException("Key 'fname' is missing in data array or is not a string");
        }

        $fname = new \HGEntity\VO\Generic\Name($data['fname']);

        if (! isset($data['lname']) || ! \is_string($data['lname'])) {
            throw new \InvalidArgumentException("Key 'lname' is missing in data array or is not a string");
        }

        $lname = new \HGEntity\VO\Generic\Name($data['lname']);

        if (! isset($data['email']) || ! \is_array($data['email'])) {
            throw new \InvalidArgumentException("Key 'email' is missing in data array or is not an array");
        }
        
        $email = [];

        foreach ($data['email'] as $__value) {
            if (! \is_HGEntity\VO\Generic\EmailAddress($__value)) {
                throw new \InvalidArgumentException("Key 'email' in data array or is not an array of HGEntity\VO\Generic\EmailAddress");
            }

            $email[] = $__value;
        }

        if (! isset($data['mobile']) || ! \is_array($data['mobile'])) {
            throw new \InvalidArgumentException("Key 'mobile' is missing in data array or is not an array");
        }
        
        $mobile = [];

        foreach ($data['mobile'] as $__value) {
            if (! \is_HGEntity\VO\Generic\Phone($__value)) {
                throw new \InvalidArgumentException("Key 'mobile' in data array or is not an array of HGEntity\VO\Generic\Phone");
            }

            $mobile[] = $__value;
        }

        if (! isset($data['phone']) || ! \is_array($data['phone'])) {
            throw new \InvalidArgumentException("Key 'phone' is missing in data array or is not an array");
        }
        
        $phone = [];

        foreach ($data['phone'] as $__value) {
            if (! \is_HGEntity\VO\Generic\Phone($__value)) {
                throw new \InvalidArgumentException("Key 'phone' in data array or is not an array of HGEntity\VO\Generic\Phone");
            }

            $phone[] = $__value;
        }

        if (! isset($data['dob']) || ! \is_string($data['dob'])) {
            throw new \InvalidArgumentException("Key 'dob' is missing in data array or is not a string");
        }

        $dob = new \HGEntity\VO\Generic\DateOfBirth($data['dob']);

        if (! isset($data['loc']) || ! \is_string($data['loc'])) {
            throw new \InvalidArgumentException("Key 'loc' is missing in data array or is not a string");
        }

        $loc = \HGEntity\VO\User\Location::fromName($data['loc']);

        if (! isset($data['group']) || ! \is_string($data['group'])) {
            throw new \InvalidArgumentException("Key 'group' is missing in data array or is not a string");
        }

        $group = \HGEntity\VO\Generic\Group::fromName($data['group']);

        if (isset($data['avatar'])) {
            if (! \is_string($data['avatar'])) {
                throw new \InvalidArgumentException("Value for 'avatar' is not a string in data array");
            }

            $avatar = new \HGEntity\VO\User\AvatarId($data['avatar']);
        } else {
            $avatar = null;
        }

        return new self(
            $id,
            $fname,
            $lname,
            $email,
            $mobile,
            $phone,
            $dob,
            $loc,
            $group,
            $avatar
        );
    }
```
Of course code errors on such ```\is_<class>``` checks.

**Comment by prolic** [comment]
@MassiAtHG Cannot reproduce on latest master.

**Comment by MassiAtHG** [comment]
Removed 'vendor' folder and then run 'composer update' on "prolic/fpp": "dev-master".
It did checkout the following reference:
```
 - Installing prolic/fpp (dev-master c580a69): Cloning c580a696c1 from cache.
```
The problem still persist.
```
        $category = [];
        foreach ($data['category'] as $__value) {
            if (! \is_HGEntity\VO\Generic\Category($__value)) {
                throw new \InvalidArgumentException("Key 'category' in data array or is not an array of HGEntity\VO\Generic\Category");
            }
            $category[] = $__value;
        }
```

**Comment by prolic** [comment]
okay, can you send me the other definitions as well? you only posted `User`.

**Comment by prolic** [comment]
If you prefer, you can send this also via email

**Comment by MassiAtHG** [comment]
Attached are the definitions for the entities.
[fpp-defs.zip](https://github.com/prolic/fpp/files/2572300/fpp-defs.zip)

**Comment by MassiAtHG** [comment]
Of course, composer.json has the PSR4 autoloading directive set up as follow:
"autoload": {
        "psr-4": {
            "HGEntity\\": "HGEntity/"
        }
    }

**Comment by prolic** [comment]
Ok I try to check when I have a little time :)
Sascha-Oliver Prolic


Am Mo., 12. Nov. 2018 um 22:46 Uhr schrieb MassiAtHG <
notifications@github.com>:

**Comment by MassiAtHG** [comment]
Since you reopened it, I tried to see if I could come up with an idea for a fix. I am not sure I am covering all the possible options properly.
The resulting code I see in my test project seems OK though (attached).
[fpp-bug-test.zip](https://github.com/prolic/fpp/files/2732449/fpp-bug-test.zip)

**Comment by MassiAtHG** [comment]
I have changed the logic of the solution in the latest commit, to be more similar to the code you wrote.
The amendment from the original code are done only on 'fromArray' builder for lists deriving from string. There may be the need to look also into the rest of the options (type and derivings) for all lists definitions. Tests are fine.

**Comment by MassiAtHG** [comment]
Found the cause of the issue, due to my deriving definitions that would not have 'FromString' or 'FromArray' (but only 'ToString' or 'ToArray') for entities on which other entities depend from. With that addition the builder (i.e. buildFromArrayBody ) can enter the correct 'switch' deriving  option.

I believe the issue can be closed, sorry if I caused you waste of time and energy.
Personally I immensely appreciate your work (and the Prooph team's).
